### PR TITLE
PARQUET-457: Verify page deserialization for GZIP and SNAPPY codecs, related refactoring

### PR DESCRIPTION
**Pull request by wesm**
This also restores passing on user's `CMAKE_CXX_FLAGS`, which had unfortunately led some compiler warnings to creep into our build.

### CODE REVIEW

**Comment by wesm** [comment]
@asandryh @majetideepak this patch is ready for review

**Comment by wesm** [comment]
rebased so I can see a coverage report

File: src/parquet/column/page.h [hunk 1]
```diff
@@ -93,13 +94,26 @@ class DataPage : public Page {
...
+
  private:
   int32_t num_values_;
   Encoding::type encoding_;
   Encoding::type definition_level_encoding_;
   Encoding::type repetition_level_encoding_;
 
-  // TODO(wesm): parquet::DataPageHeader.statistics
+  // So max/min can be populated privately
+  friend class SerializedPageReader;
```
Comment (majetideepak) [inline src/parquet/column/page.h]: Why do we need `friend` here?

Comment (majetideepak) [inline src/parquet/column/page.h]: Nevermind, I got this.

**Comment by majetideepak** [comment]
+1 LGTM

File: src/parquet/column/page.h (hunk 1 above)
Comment (wesm) [inline src/parquet/column/page.h]: I wasn't prepared to work on the public API for handling page statistics, and this was the minimum effort that would preserve the existing unit tests through the refactor.

**Comment by wesm** [comment]
There's a gzip memory leak, will add a commit to fix.

**Comment by wesm** [comment]
OK, now this can be merged.

**Comment by majetideepak** [comment]
We could run Valgrind at some point.

**Comment by wesm** [comment]
Valgrind is already being run in the test suite -- the memory leak I saw was transient. I haven't been able to reproduce it, but the fix here is the right one (looked at Impala's gzip code for guidance).

**Comment by wesm** [comment]
I just squashed to simplify patch maintenance and rebasing downstream

**Comment by majetideepak** [comment]
I meant run Valgrind for the entire parquet-cpp. Should I open a new JIRA to track this? We can add a check to travis ci as well.
I saw this output on my machine. 
==32506== HEAP SUMMARY:
==32506==     in use at exit: 5,436 bytes in 52 blocks
==32506==   total heap usage: 766 allocs, 714 frees, 135,749 bytes allocated
... 52 not-freed blocks
----------------------------

**Comment by wesm** [comment]
There is a check in Travis

https://github.com/apache/parquet-cpp/blob/master/.travis.yml#L37

Here's the output from this PR. The behavior may vary depending on the platform.

```
2.39s$ valgrind --tool=memcheck --leak-check=yes --error-exitcode=1 ctest
==16267== Memcheck, a memory error detector
==16267== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==16267== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==16267== Command: ctest
==16267== 
Test project /home/travis/build/apache/parquet-cpp/parquet-build
      Start  1: public-api-test
 1/14 Test  #1: public-api-test ..................   Passed    0.07 sec
      Start  2: reader-test
 2/14 Test  #2: reader-test ......................   Passed    0.04 sec
      Start  3: column-reader-test
 3/14 Test  #3: column-reader-test ...............   Passed    0.04 sec
      Start  4: levels-test
 4/14 Test  #4: levels-test ......................   Passed    0.04 sec
      Start  5: codec-test
 5/14 Test  #5: codec-test .......................   Passed    0.08 sec
      Start  6: plain-encoding-test
 6/14 Test  #6: plain-encoding-test ..............   Passed    0.06 sec
      Start  7: file-deserialize-test
 7/14 Test  #7: file-deserialize-test ............   Passed    0.04 sec
      Start  8: schema-converter-test
 8/14 Test  #8: schema-converter-test ............   Passed    0.04 sec
      Start  9: schema-descriptor-test
 9/14 Test  #9: schema-descriptor-test ...........   Passed    0.04 sec
      Start 10: schema-printer-test
10/14 Test #10: schema-printer-test ..............   Passed    0.04 sec
      Start 11: schema-types-test
11/14 Test #11: schema-types-test ................   Passed    0.04 sec
      Start 12: bit-util-test
12/14 Test #12: bit-util-test ....................   Passed    0.04 sec
      Start 13: output-test
13/14 Test #13: output-test ......................   Passed    0.04 sec
      Start 14: rle-test
14/14 Test #14: rle-test .........................   Passed    0.10 sec
100% tests passed, 0 tests failed out of 14
Total Test time (real) =   1.08 sec
==16267== 
==16267== HEAP SUMMARY:
==16267==     in use at exit: 72,704 bytes in 1 blocks
==16267==   total heap usage: 15,475 allocs, 15,474 frees, 2,168,334 bytes allocated
==16267== 
==16267== LEAK SUMMARY:
==16267==    definitely lost: 0 bytes in 0 blocks
==16267==    indirectly lost: 0 bytes in 0 blocks
==16267==      possibly lost: 0 bytes in 0 blocks
==16267==    still reachable: 72,704 bytes in 1 blocks
==16267==         suppressed: 0 bytes in 0 blocks
==16267== Reachable blocks (those to which a pointer was found) are not shown.
==16267== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==16267== 
==16267== For counts of detected and suppressed errors, rerun with: -v
==16267== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

We should probably treat still-reachable blocks as leaks (which do not yield error valgrind status) -- feel free to create JIRAs as you find reproducible memory leaks.

**Comment by majetideepak** [comment]
Interesting! I will investigate more on this later and add JIRAs if needed.

**Comment by julienledem** [comment]
+1

**Pull request closed without merging**